simulation/traffic_simulator/src/api/api.cpp

```diff
@@ -232,19 +232,41 @@ bool API::updateTrafficLightsInSim()
   return simulation_api_schema::UpdateTrafficLightsResponse().result().success();
 }
 
+bool API::updateNonEgoEntitiesStatusInSim()
+{
+  simulation_api_schema::UpdateEntityStatusRequest req;
+  req.set_npc_logic_started(entity_manager_ptr_->isNpcLogicStarted());
+
+  for (const auto & entity_name : entity_manager_ptr_->getEntityNames()) {
+    if (entity_manager_ptr_->isEgo(entity_name)) continue;
+    auto status = entity_manager_ptr_->getEntityStatus(entity_name);
+    simulation_interface::toProto(static_cast<EntityStatus>(status), *req.add_status());
+  }
+  return zeromq_client_.call(req).result().success();
+}
+
 std::optional<CanonicalizedEntityStatus> API::updateEntityStatusInSim(
   const std::string & entity_name, const CanonicalizedEntityStatus & status)
 {
+  constexpr int single_status_update = 1;
+
   simulation_api_schema::UpdateEntityStatusRequest req;
-  simulation_api_schema::EntityStatus proto;
+  req.set_npc_logic_started(entity_manager_ptr_->isNpcLogicStarted());
   auto status_non_canonicalized = static_cast<EntityStatus>(status);
   status_non_canonicalized.name = entity_name;
-  simulation_interface::toProto(status_non_canonicalized, *req.mutable_status());
-  req.set_npc_logic_started(entity_manager_ptr_->isNpcLogicStarted());
-  if (auto res = zeromq_client_.call(req); res.result().success()) {
-    simulation_interface::toMsg(res.status().pose(), status_non_canonicalized.pose);
+  simulation_interface::toProto(status_non_canonicalized, *req.add_status());
+
+  simulation_api_schema::UpdateEntityStatusResponse res;
+  // The method updates a single entity status, so a single status is expected in response
+  if (auto res = zeromq_client_.call(req);
+      res.result().success() && req.status_size() == single_status_update) {
```

Comment:
When we use simple_sensor_simulator, `req.status_size()` should be 2, but the simulator such as AWSIM has a dynamics engine, so such simulators can change non-ego entity status.
So, we should check all updated entity status.

Reply:
@hakuturu583 
Do you mean that even though there is only one status to update in the request, we can receive several statuses in the response (because they could change independently of the requests)?

Reply:
UpdateEntityStatusRequest send status of multiple entities.
I have determined that some simulator types may want to update EntityStatus during frame updates, independently of the traffic_simulator logic.
So, not only Ego entity but also non-Ego entity status can be updated by sensor simulator.

Reply:
@hakuturu583 
Thank you for your answer, please confirm which version best corresponds to the expected implementation:

#### Current
![current drawio](https://github.com/tier4/scenario_simulator_v2/assets/121798334/ede772b6-17cd-4803-b328-ab47c3b0a99f)

#### First proposal
![processed drawio](https://github.com/tier4/scenario_simulator_v2/assets/121798334/697678f3-b41f-44d9-9220-24dce7cd37a2)

#### Second proposal
![multiple drawio](https://github.com/tier4/scenario_simulator_v2/assets/121798334/99508b07-41a3-413d-8343-7fa13b76c78c)

### Received `UpdatedEntityStatus`
As I understand in both proposals, it is necessary to implement the processing of the received status (non-Ego entities) on the ss2 side.

Please let me know if the received status should be simply rewritten to `EntityManager`, or if a check should be implemented: if the received status is equal to the requested one.  Such a check seems to me to be essential.
I mean checking if the values in the received `Pose` are equal to the requested one - taking into account the `epsilon` value, e.g. `0.001` (does `ss2` have a parameter that can be used in such a check?).

Also, let me know what is the expected behavior when the expected values are not set - should an exception be thrown?

Reply:
![Screenshot from 2023-08-25 11-14-10](https://github.com/tier4/scenario_simulator_v2/assets/10348912/b4dc2e2a-0651-41a0-9dff-018a4f443150)

I think this design is better because of overhead of the communication is minimum.

I think only updated entity status in `simple_sensor_simulator/AWSIM` should be packed into response.
If the requested status does not exists in response, (it can be checked by data field [name](https://github.com/tier4/scenario_simulator_v2/blob/408b00a4ffd3a8066cc777b72d40404f38fee141/simulation/simulation_interface/proto/traffic_simulator_msgs.proto#L75).) it means `No Operation` so do not update entity status in traffic_simulator.

Reply:
[Currently, non-ego entity does not it's status by returned response.](https://github.com/tier4/scenario_simulator_v2/blob/290ff1b9c2bab79b613ceb2aaa3acb0a1c7c55b6/simulation/traffic_simulator/src/api/api.cpp#L235-L246)
So, it should be also implemented.

Reply:
Currently, you implement update ego entity status before updating non-ego entity status.
I think we can do it at the same time.
Is there some problem about this plan?